### ISSUE TITLE
Parallel config seems to be broken

### ISSUE THREAD

**Issue by dennisameling** [acrobat/subtree-splitter#1]
First of all, thank you for creating this GitHub Action! Really appreciate that 🚀 

We implemented it on https://github.com/mautic/mautic but ran into an issue with timeouts. More specifically:

**Scenario 1: run all configs in parallel**

GH Actions config: https://github.com/mautic/mautic/blob/86b4c17918212d075dcc7d6bbe95f7d702973f18/.github/workflows/split-monorepo-in-multi-repo.yml
Run: https://github.com/mautic/mautic/runs/4162757834?check_suite_focus=true

This leads to the following error:

```
┌─────────┬──────────────────────────────┬──────────────────────────────────────┬─────────────────────────────────────────────────────┐
│ (index) │             name             │              directory               │                       target                        │
├─────────┼──────────────────────────────┼──────────────────────────────────────┼─────────────────────────────────────────────────────┤
│    0    │          'core-lib'          │                'app'                 │        'git@github.com:mautic/core-lib.git'         │
│    1    │   'GrapesJsBuilderBundle'    │   'plugins/GrapesJsBuilderBundle'    │ 'git@github.com:mautic/plugin-grapesjs-builder.git' │
│    2    │     'MauticCitrixBundle'     │     'plugins/MauticCitrixBundle'     │      'git@github.com:mautic/plugin-citrix.git'      │
│    3    │    'MauticClearbitBundle'    │    'plugins/MauticClearbitBundle'    │     'git@github.com:mautic/plugin-clearbit.git'     │
│    4    │  'MauticCloudStorageBundle'  │  'plugins/MauticCloudStorageBundle'  │   'git@github.com:mautic/plugin-cloudstorage.git'   │
│    5    │      'MauticCrmBundle'       │      'plugins/MauticCrmBundle'       │       'git@github.com:mautic/plugin-crm.git'        │
│    6    │ 'MauticEmailMarketingBundle' │ 'plugins/MauticEmailMarketingBundle' │  'git@github.com:mautic/plugin-emailmarketing.git'  │
│    7    │     'MauticFocusBundle'      │     'plugins/MauticFocusBundle'      │      'git@github.com:mautic/plugin-focus.git'       │
│    8    │  'MauticFullContactBundle'   │  'plugins/MauticFullContactBundle'   │   'git@github.com:mautic/plugin-fullcontact.git'    │
│    9    │     'MauticGmailBundle'      │     'plugins/MauticGmailBundle'      │      'git@github.com:mautic/plugin-gmail.git'       │
│   10    │    'MauticOutlookBundle'     │    'plugins/MauticOutlookBundle'     │     'git@github.com:mautic/plugin-outlook.git'      │
│   11    │     'MauticSocialBundle'     │     'plugins/MauticSocialBundle'     │      'git@github.com:mautic/plugin-social.git'      │
│   12    │   'MauticTagManagerBundle'   │   'plugins/MauticTagManagerBundle'   │    'git@github.com:mautic/plugin-tagmanager.git'    │
│   13    │     'MauticZapierBundle'     │     'plugins/MauticZapierBundle'     │      'git@github.com:mautic/plugin-zapier.git'      │
│   14    │        'theme-aurora'        │           'themes/aurora'            │      'git@github.com:mautic/theme-aurora.git'       │
│   15    │        'theme-blank'         │            'themes/blank'            │       'git@github.com:mautic/theme-blank.git'       │
│   16    │        'theme-brienz'        │           'themes/brienz'            │      'git@github.com:mautic/theme-brienz.git'       │
│   17    │        'theme-cards'         │            'themes/cards'            │       'git@github.com:mautic/theme-cards.git'       │
│   18    │        'theme-coffee'        │           'themes/coffee'            │      'git@github.com:mautic/theme-coffee.git'       │
│   19    │      'theme-confirmme'       │          'themes/confirmme'          │     'git@github.com:mautic/theme-confirmme.git'     │
│   20    │     'theme-fresh-center'     │        'themes/fresh-center'         │   'git@github.com:mautic/theme-fresh-center.git'    │
│   21    │     'theme-fresh-fixed'      │         'themes/fresh-fixed'         │    'git@github.com:mautic/theme-fresh-fixed.git'    │
│   22    │      'theme-fresh-left'      │         'themes/fresh-left'          │    'git@github.com:mautic/theme-fresh-left.git'     │
│   23    │      'theme-fresh-wide'      │         'themes/fresh-wide'          │    'git@github.com:mautic/theme-fresh-wide.git'     │
│   24    │       'theme-goldstar'       │          'themes/goldstar'           │     'git@github.com:mautic/theme-goldstar.git'      │
│   25    │        'theme-mauve'         │            'themes/Mauve'            │       'git@github.com:mautic/theme-mauve.git'       │
│   26    │        'theme-nature'        │           'themes/nature'            │      'git@github.com:mautic/theme-nature.git'       │
│   27    │      'theme-neopolitan'      │         'themes/neopolitan'          │    'git@github.com:mautic/theme-neopolitan.git'     │
│   28    │        'theme-oxygen'        │           'themes/oxygen'            │      'git@github.com:mautic/theme-oxygen.git'       │
│   29    │       'theme-paprika'        │           'themes/paprika'           │      'git@github.com:mautic/theme-paprika.git'      │
│   30    │       'theme-skyline'        │           'themes/sparse'            │      'git@github.com:mautic/theme-sparse.git'       │
│   31    │        'theme-sunday'        │           'themes/sunday'            │      'git@github.com:mautic/theme-sunday.git'       │
│   32    │    'theme-trulypersonal'     │        'themes/trulypersonal'        │   'git@github.com:mautic/theme-trulypersonal.git'   │
│   33    │       'theme-vibrant'        │           'themes/vibrant'           │      'git@github.com:mautic/theme-vibrant.git'      │
└─────────┴──────────────────────────────┴──────────────────────────────────────┴─────────────────────────────────────────────────────┘
/usr/bin/git remote add core-lib git@github.com:mautic/core-lib.git
/usr/bin/git remote add GrapesJsBuilderBundle git@github.com:mautic/plugin-grapesjs-builder.git
/usr/bin/git remote add MauticCitrixBundle git@github.com:mautic/plugin-citrix.git
/usr/bin/git remote add MauticClearbitBundle git@github.com:mautic/plugin-clearbit.git
/usr/bin/git remote add MauticCloudStorageBundle git@github.com:mautic/plugin-cloudstorage.git
/usr/bin/git remote add MauticCrmBundle git@github.com:mautic/plugin-crm.git
/usr/bin/git remote add MauticEmailMarketingBundle git@github.com:mautic/plugin-emailmarketing.git
/usr/bin/git remote add MauticFocusBundle git@github.com:mautic/plugin-focus.git
/usr/bin/git remote add MauticFullContactBundle git@github.com:mautic/plugin-fullcontact.git
/usr/bin/git remote add MauticGmailBundle git@github.com:mautic/plugin-gmail.git
/usr/bin/git remote add MauticOutlookBundle git@github.com:mautic/plugin-outlook.git
/usr/bin/git remote add MauticSocialBundle git@github.com:mautic/plugin-social.git
/usr/bin/git remote add MauticTagManagerBundle git@github.com:mautic/plugin-tagmanager.git
/usr/bin/git remote add MauticZapierBundle git@github.com:mautic/plugin-zapier.git
/usr/bin/git remote add theme-aurora git@github.com:mautic/theme-aurora.git
/usr/bin/git remote add theme-blank git@github.com:mautic/theme-blank.git
/usr/bin/git remote add theme-brienz git@github.com:mautic/theme-brienz.git
/usr/bin/git remote add theme-cards git@github.com:mautic/theme-cards.git
/usr/bin/git remote add theme-coffee git@github.com:mautic/theme-coffee.git
/usr/bin/git remote add theme-confirmme git@github.com:mautic/theme-confirmme.git
/usr/bin/git remote add theme-fresh-center git@github.com:mautic/theme-fresh-center.git
/usr/bin/git remote add theme-fresh-fixed git@github.com:mautic/theme-fresh-fixed.git
/usr/bin/git remote add theme-fresh-left git@github.com:mautic/theme-fresh-left.git
/usr/bin/git remote add theme-fresh-wide git@github.com:mautic/theme-fresh-wide.git
/usr/bin/git remote add theme-goldstar git@github.com:mautic/theme-goldstar.git
/usr/bin/git remote add theme-mauve git@github.com:mautic/theme-mauve.git
/usr/bin/git remote add theme-nature git@github.com:mautic/theme-nature.git
/usr/bin/git remote add theme-neopolitan git@github.com:mautic/theme-neopolitan.git
/usr/bin/git remote add theme-oxygen git@github.com:mautic/theme-oxygen.git
/usr/bin/git remote add theme-paprika git@github.com:mautic/theme-paprika.git
/usr/bin/git remote add theme-skyline git@github.com:mautic/theme-sparse.git
/usr/bin/git remote add theme-sunday git@github.com:mautic/theme-sunday.git
/usr/bin/git remote add theme-trulypersonal git@github.com:mautic/theme-trulypersonal.git
/usr/bin/git remote add theme-vibrant git@github.com:mautic/theme-vibrant.git
/home/runner/work/mautic/mautic/splitsh --prefix=themes/Mauve --origin=origin/4.x
/home/runner/work/mautic/mautic/splitsh --prefix=themes/goldstar --origin=origin/4.x
/home/runner/work/mautic/mautic/splitsh --prefix=themes/fresh-wide --origin=origin/4.x
/home/runner/work/mautic/mautic/splitsh --prefix=themes/fresh-left --origin=origin/4.x
/home/runner/work/mautic/mautic/splitsh --prefix=themes/fresh-fixed --origin=origin/4.x
/home/runner/work/mautic/mautic/splitsh --prefix=themes/fresh-center --origin=origin/4.x
/home/runner/work/mautic/mautic/splitsh --prefix=themes/confirmme --origin=origin/4.x
/home/runner/work/mautic/mautic/splitsh --prefix=themes/coffee --origin=origin/4.x
/home/runner/work/mautic/mautic/splitsh --prefix=themes/cards --origin=origin/4.x
/home/runner/work/mautic/mautic/splitsh --prefix=themes/brienz --origin=origin/4.x
/home/runner/work/mautic/mautic/splitsh --prefix=themes/blank --origin=origin/4.x
/home/runner/work/mautic/mautic/splitsh --prefix=themes/aurora --origin=origin/4.x
/home/runner/work/mautic/mautic/splitsh --prefix=plugins/MauticZapierBundle --origin=origin/4.x
/home/runner/work/mautic/mautic/splitsh --prefix=plugins/MauticTagManagerBundle --origin=origin/4.x
/home/runner/work/mautic/mautic/splitsh --prefix=plugins/MauticSocialBundle --origin=origin/4.x
/home/runner/work/mautic/mautic/splitsh --prefix=plugins/MauticOutlookBundle --origin=origin/4.x
/home/runner/work/mautic/mautic/splitsh --prefix=plugins/MauticGmailBundle --origin=origin/4.x
/home/runner/work/mautic/mautic/splitsh --prefix=plugins/MauticFullContactBundle --origin=origin/4.x
/home/runner/work/mautic/mautic/splitsh --prefix=plugins/MauticFocusBundle --origin=origin/4.x
/home/runner/work/mautic/mautic/splitsh --prefix=plugins/MauticEmailMarketingBundle --origin=origin/4.x
/home/runner/work/mautic/mautic/splitsh --prefix=plugins/MauticCrmBundle --origin=origin/4.x
/home/runner/work/mautic/mautic/splitsh --prefix=plugins/MauticCloudStorageBundle --origin=origin/4.x
/home/runner/work/mautic/mautic/splitsh --prefix=plugins/MauticClearbitBundle --origin=origin/4.x
/home/runner/work/mautic/mautic/splitsh --prefix=plugins/MauticCitrixBundle --origin=origin/4.x
/home/runner/work/mautic/mautic/splitsh --prefix=plugins/GrapesJsBuilderBundle --origin=origin/4.x
/home/runner/work/mautic/mautic/splitsh --prefix=app --origin=origin/4.x
error: could not lock config file .git/config: File exists
fatal: could not set 'remote.theme-nature.url' to 'git@github.com:mautic/theme-nature.git'
Error: Error: The process '/usr/bin/git' failed with exit code 128
/home/runner/work/mautic/mautic/splitsh --prefix=themes/vibrant --origin=origin/4.x
/home/runner/work/mautic/mautic/splitsh --prefix=themes/trulypersonal --origin=origin/4.x
/home/runner/work/mautic/mautic/splitsh --prefix=themes/sunday --origin=origin/4.x
/home/runner/work/mautic/mautic/splitsh --prefix=themes/sparse --origin=origin/4.x
/home/runner/work/mautic/mautic/splitsh --prefix=themes/paprika --origin=origin/4.x
/home/runner/work/mautic/mautic/splitsh --prefix=themes/oxygen --origin=origin/4.x
/home/runner/work/mautic/mautic/splitsh --prefix=themes/neopolitan --origin=origin/4.x
timeout
timeout
timeout
timeout
timeout
timeout
timeout
timeout
timeout
timeout
timeout
timeout
timeout
timeout
timeout
timeout
timeout
timeout
timeout
timeout
timeout
timeout
timeout
timeout
timeout
timeout
timeout
timeout
timeout
timeout
timeout
timeout
52 commits created, 23250 commits traversed, in 16.044s
642aee92a8a1fd6d5344cd82516851d0087bd5eb
/usr/bin/git push theme-mauve 642aee92a8a1fd6d5344cd82516851d0087bd5eb:refs/heads/4.x -f
To https://github.com/mautic/theme-mauve.git
 + 301b8e82b6...642aee92a8 642aee92a8a1fd6d5344cd82516851d0087bd5eb -> 4.x (forced update)
```

**Scenario 2: run all configs in separate jobs**

We [applied a workaround today](https://github.com/mautic/mautic/pull/10599) by running all configs separately - this works without problems for us, but eventually we'd like to have the single config file for all repos again.

GH Actions config: https://github.com/mautic/mautic/blob/c9c6465cfe097ac10f89f043e7fc79101fa2053a/.github/workflows/split-monorepo-in-multi-repo.yml
Run: https://github.com/mautic/mautic/actions/runs/1443942846

With this config, all is OK for now ✅ 

Do you have any idea what could be causing the error/timeouts when we run all configs at the same time? Thanks in advance! 😊

**Comment by acrobat** [comment]
Hi @dennisameling, thanks for the bug report! I will try to reproduce it and see what is causing this isssue!

**Comment by acrobat** [comment]
Hi @dennisameling, sorry for the delay but I was able to reproduce this partially. There could indeed happen a race condition in adding the remotes as they "edit" the same config file. The timeouts itself I was not able to reproduce and seem to be a temporary connection issue on github actions itself.

Can you try `acrobat/subtree-splitter@concurrency-fix` in your project and see if the problem is resolved? Thanks!

**Comment by dennisameling** [comment]
Hi @acrobat, really sorry for the late reply here. Just tested with the latest changes and it's still timing out unfortunately. Do you have any ideas on what this could be?

CI run: https://github.com/mautic/mautic/runs/5373193520?check_suite_focus=true
Config: https://github.com/mautic/mautic/pull/10910

**Comment by acrobat** [comment]
hi @dennisameling! I seems that the amount of splits and with it the amount of splitsh processes running are too much for the github actions runners. Let me see what I can do to split up the load more, I will ping you if I have a solution that you can test on that PR!

**Comment by acrobat** [comment]
@dennisameling can you try the patch in #3? This will process all configured splits in batches of 4, which hopefully should spread the load more

**Comment by dennisameling** [comment]
Thanks for looking into this! It's still timing out unfortunately: https://github.com/mautic/mautic/runs/5413138732?check_suite_focus=true

Note that the timeouts already show up after ~10 seconds. My gut says that it still tries to kick off everything at once instead of 4 at a time.

Thanks again - hope our testing helps a bit! 🙏🏼

**Comment by acrobat** [comment]
You are right, on the main branch the action code gets rebuild automatically but on others not. So you were actually running the old version still, I've re-compiled a new version that you can check.

There should now also be a line telling which batch is being processed to help us debug it. `Processing batch 1/x`

**Comment by dennisameling** [comment]
Here's another fresh run: https://github.com/mautic/mautic/runs/5431857880?check_suite_focus=true

It seems that there can only be 1 job running at a time - 3/4 time out almost immediately and only 1 `splitsh` process actually runs.

Thanks a lot for your help so far in debugging/resolving this issue!

**Comment by Zales0123** [comment]
Hello! 

Thanks a lot for this library as well, @acrobat :)

I'm trying to use it in [Sylius](https://github.com/Sylius/Sylius), as we're already using the split mechanism and I would like to automatize it. However, I've encountered the same problem as described in this issue, so I'm bumping it up, as an important one 🚀 For now, I believe I will use the workaround used in mautic and described by @dennisameling (thank you for this!), but would be great to have it fixed here in the action itself 🖖 

Have a good day!

**Comment by acrobat** [comment]
Sorry for the late answer but I've implemented a feature to workaround this issue. For small repo's (low commits) the current setup works but for larger repo's I've added the option to process everything in batches. This seems to work for me in [Kunstmaan/KunstmaanBundlesCMS](https://github.com/Kunstmaan/KunstmaanBundlesCMS) as I didn't have any timeouts and the whole process is still fast enough for me.

https://github.com/Kunstmaan/KunstmaanBundlesCMS/blob/d25302a37a006c8e70a44e3df3f89f40cbfb64f2/.github/workflows/subtree-split.yml#L42-L46

So if you can test this feature/fix and see if it solves your problem. Use `acrobat/subtree-splitter@batch-process-splits` and setup the `batch-size` config. 

See #3

**Comment by acrobat** [comment]
I've merged the PR mentioned above (it's included in the [1.1.0](https://github.com/acrobat/subtree-splitter/releases/tag/v1.1.0) release) as we've been using it without issues for quite some time on the kunstmaan cms repo. Please reopen the issue or create a new one if you still experience issues with the new batchSize option.
